File: DKPCounter/src/org/henrya/ronin/dkpcounter/Boss.java
package org.henrya.ronin.dkpcounter;

/**
 * A class representative of each boss that points are awarded for
 * @author Henry Anderson
 *
 */
public enum Boss {
	/* DL bosses */
	DL_155(new String[] {"155", "spider", "spidey", "ulrob"}, 0, new int[]{2, 5, 10}),
	DL_160(new String[] {"160", "priest"}, 0, new int[] {2, 5, 10}),
	DL_165(new String[] {"165", "king"}, 0, new int[] {2, 5, 10}),
	DL_170(new String[] {"170", "sreng", "bolg"}, 5, new int[] {2, 5, 10}),
	DL_180(new String[] {"180", "snorri", "snor", "snorr", "troll"}, 5, new int[] {2, 5, 10}),

	/* EDL bosses */
	EDL_185(new String[] {"185", "dog", "doggy", "dogg", "ifryn"}, 0, new int[] {5, 10, 15}),
	EDL_190(new String[] {"190", "skath", "magister"}, 0, new int[] {5, 10, 15}),
	EDL_195(new String[] {"195", "gron", "general"}, 0, new int[] {5, 10, 15}),
	EDL_200(new String[] {"200", "krother", "imp"}, 0, new int[] {5, 10, 15}),
	EDL_205(new String[] {"205", "cragskor", "dragon"}, 5, new int[] {5, 10, 15}),
	EDL_210(new String[] {"210", "revenant", "anguish"}, 5, new int[] {5, 10, 15}),
	EDL_215(new String[] {"215", "unox", "eye"}, 5, new int[] {5, 10, 15}),
	
	/* Raids */
	AGGRAGOTH(new String[] {"aggragoth", "aggy", "agg"}, 0, new int[] {0, 0, 10}),
	HRUNGNIR(new String[] {"hrungnir", "hrung", "hippo"}, 0, new int[] {0, 0, 15}),
	MORDRIS(new String[] {"mordris", "mordy", "mord"}, 0, new int[] {0, 0, 20}),
	EFNISIEN(new String[] {"efnisien", "efni", "necromancer", "necro"}, 0, new int[] {0, 0, 25}),
	
	PROTEUS_BASE(new String[] {"proteusbase", "proteus_base", "protb", "base"}, 0, new int[] {0, 0, 60}),
	PROTEUS_PRIME(new String[] {"proteusprime", "proteus_prime", "protp", "prime"}, 0, new int[] {0, 0, 80}),
	GELEBRON(new String[] {"gelebron", "gele", "gel"}, 0, new int[] {0, 0, 100}),
	BLOODTHORN(new String[] {"bloodthorn", "bt", "blood", "thorn"}, 0, new int[] {0, 0, 150}),

	/* Arcane and Gardens extra */
	RING_BOSS(new String[] {"ringboss", "ring_boss", "rb", "rbs", "ring"}, 0, new int[] {0, 20, 40}),
	FACTIONS(new String[] {"factions", "faction", "lich", "reaver"}, 0, new int[] {0, 0, 30}),
	BLOODTHORN_ROOT(new String[] {"bloodthornroot", "bloodthorn_root", "btroot", "bt_root", "root"}, 0, new int[] {0, 10, 0});

	private String[] aliases;
	private int campPoints;
	private int[] killPoints;

	/** 
	 * Constructs a new Boss instance
	 * @param aliases The names it can be referred to as
	 * @param campPoints The amount of camp points to award
	 * @param killPoints The amount of kill points to award
	 */
	Boss(String[] aliases, int campPoints, int[] killPoints) {
		this.aliases = aliases;
		this.campPoints = campPoints;
		this.killPoints = killPoints;
	}
	
	/**
	 * Returns the aliases of the boss
	 * @return An array of aliases
	 */
	public String[] getAliases() {
		return this.aliases;
	}
	
	/**
	 * Returns the camp points to award
	 * @return The camp points
	 */
	public int getCampPoints() {
		return this.campPoints;
	}
	
	/**
	 * Returns the kill points to award in an array
	 * @return The kill points
	 */
	public int[] getKillPoints() {
		return this.getKillPoints();
	}
	
	/**
	 * Returns the kill points to award based on the amount of stars
	 * @param stars The stars
	 * @return The kill points
	 */
	public int getKillPoints(int stars) {
		if(stars == 4) {
			return this.killPoints[0];
		}
		else if(stars == 5) {
			return this.killPoints[1];
		}
		else if(stars == 6) { 
			return this.killPoints[2];
		}
		return 0;
	}
	
	/**
	 * Returns a Boss instance based on its name
	 * @param name The name of the boss
	 * @return The boss
	 */
	public static Boss getBoss(String name) {
		for(Boss boss : Boss.values()) {
			for(String alias : boss.getAliases()) {
				if(name.equalsIgnoreCase(alias)) {
					return boss;
				}
			}
		}
		return null;
	}
}

File: docs/count/boss.js
const Boss = {

    /** DL bosses **/
    DL_155: {
        aliases: ["155", "spider", "spidey", "ulrob"],
        points: {
            camp: 0,
            kill4: 2,
            kill5: 5,
            kill6: 10,
        }
    },
    DL_160: {
        aliases: ["160", "priest"],
        points: {
            camp: 0,
            kill4: 2,
            kill5: 5,
            kill6: 10,
        }
    },
    DL_165: {
        aliases: ["165", "king"],
        points: {
            camp: 0,
            kill4: 2,
            kill5: 5,
            kill6: 10,
        }
    },
    DL_170: {
        aliases: ["170", "sreng", "bolg"],
        points: {
            camp: 5,
            kill4: 2,
            kill5: 5,
            kill6: 10,
        }
    },
    DL_180: {
        aliases: ["180", "snorri", "snor", "snorr", "troll"],
        points: {
            camp: 5,
            kill4: 2,
            kill5: 5,
            kill6: 10,
        }
    },

    /** EDL bosses **/
    EDL_185: {
        aliases: ["185", "dog", "doggy", "dogg", "ifryn"],
        points: {
            camp: 0,
            kill4: 5,
            kill5: 10,
            kill6: 15,
        }
    },
    EDL_190: {
        aliases: ["190", "skath", "magister"],
        points: {
            camp: 0,
            kill4: 5,
            kill5: 15,
            kill6: 20,
        }
    },
    EDL_195: {
        aliases: ["195", "gron", "general"],
        points: {
            camp: 0,
            kill4: 5,
            kill5: 15,
            kill6: 20,
        }
    },
    EDL_200: {
        aliases: ["200", "krother", "imp"],
        points: {
            camp: 0,
            kill4: 5,
            kill5: 15,
            kill6: 20,
        }
    },
    EDL_205: {
        aliases: ["205", "cragskor", "dragon"],
        points: {
            camp: 0,
            kill4: 5,
            kill5: 10,
            kill6: 15,
        }
    },
    EDL_210: {
        aliases: ["210", "revenant", "anguish"],
        points: {
            camp: 10,
            kill4: 5,
            kill5: 15,
            kill6: 20,
        }
    },
    EDL_215: {
        aliases: ["215", "unox", "eye"],
        points: {
            camp: 10,
            kill4: 5,
            kill5: 15,
            kill6: 20,
        }
    },

    /** Raids **/
    AGGRAGOTH: {
        aliases: ["aggragoth", "aggy", "agg"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 10,
        }
    },
    HRUNGNIR: {
        aliases: ["hrungnir", "hrung", "hippo"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 20,
        }
    },
    MORDRIS: {
        aliases: ["mordris", "mordy", "mord", "mordi"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 35,
        }
    },
    EFNISIEN: {
        aliases: ["efnisien", "efni", "necromancer", "necro"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 40,
        }
    },

    PROTEUS_BASE: {
        aliases: ["proteusbase", "proteus_base", "protb", "base"],
        points: {
            camp: 5,
            kill4: 0,
            kill5: 0,
            kill6: 60,
        }
    },
    PROTEUS_PRIME: {
        aliases: ["proteusprime", "proteus_prime", "protp", "prime"],
        points: {
            camp: 5,
            kill4: 0,
            kill5: 0,
            kill6: 80,
        }
    },
    GELEBRON: {
        aliases: ["gelebron", "gele", "gel"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 100,
        }
    },
    BLOODTHORN: {
        aliases: ["bloodthorn", "bt", "blood", "thorn"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 150,
        }
    },
    DHIOTHU: {
        aliases: ["dhiothu", "dhio", "dino"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 200,
        }
    },
    /** Arcane and Garden extra **/
    RING_BOSS: {
        aliases: ["ringboss", "ring_boss", "rb", "rbs", "ring"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 10,
            kill6: 40,
        }
    },
    FACTIONS: {
        aliases: ["factions", "faction", "lich", "reaver"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 0,
            kill6: 30,
        }
    },
    BLOODTHORN_ROOT: {
        aliases: ["bloodthornroot", "bloodthorn_root", "btroot", "bt_root", "root"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 10,
            kill6: 0,
        }
    },

    /** Legacies **/
    LEGACY_150: {
        aliases: ["legacy150", "legacy_150", "witchfinder", "hopkins", "comrack", "akais", "hurricanes", "typhoons", "garyn", "devestator", "jalan", "paradai", "blizzardgore", "synak"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 5,
            kill6: 10,
        }
    },
    LEGACY_180: {
        aliases: ["legacy180", "legacy_180", "natan", "athatch", "garanak", "gara", "urchaid", "maelstrom", "whirlwinds", "whirlwind", "smolach", "annihilator", "skain", "osan", "blight", "tomb"],
        points: {
            camp: 0,
            kill4: 0,
            kill5: 10,
            kill6: 15,
        }
    },
};

function getKillPoints(boss, stars) {
    if(stars === 4) {
        return boss.points.kill4;
    }
    else if(stars === 5) {
        return boss.points.kill5;
    }
    else if(stars === 6) {
        return boss.points.kill6;
    }
}
function getBosses() {
    return [Boss.DL_155, Boss.DL_160, Boss.DL_165, Boss.DL_170, Boss.DL_180, Boss.EDL_185, Boss.EDL_190, Boss.EDL_195, Boss.EDL_200, Boss.EDL_205, Boss.EDL_210, Boss.EDL_215, Boss.AGGRAGOTH, Boss.HRUNGNIR, Boss.MORDRIS, Boss.EFNISIEN, Boss.PROTEUS_BASE, Boss.PROTEUS_PRIME, Boss.GELEBRON, Boss.BLOODTHORN, Boss.DHIOTHU, Boss.RING_BOSS, Boss.FACTIONS, Boss.BLOODTHORN_ROOT, Boss.LEGACY_150, Boss.LEGACY_180];
}

function getBoss(name) {
    for(let i = 0; i < getBosses().length; i++) {
        let boss = getBosses()[i];
        for(let j = 0; j < boss.aliases.length; j++) {
            if(boss.aliases[j] === name) {
                return boss;
            }
        }
    }
    return null;
}

const PointType = {
    KILL: ["kill", "killl", "kil", "kkil", "ki", "k", "kll", "kl"],
    CAMP: ["camp", "campp", "cam", "ca", "c", "cmp", "cp"],
    CAMP_KILL: ["campkill", "ckill", "campk", "campkill", "killcamp", "kcamp", "camp/kill", "kill/camp"],
};

//points here are per hour camped or per level
const BonusType = {
	LEVEL: {
        aliases: ["leveled", "level", "levelled"],
        points: 1,
    },
	CAMP: {
        aliases: ["camped"],
        points: 7,
    },
};

function getPointTypes() {
    return [PointType.CAMP, PointType.KILL, PointType.CAMP_KILL];
}

function getBonusTypes() {
    return [BonusType.LEVEL, BonusType.CAMP];
}

function getPointType(type) {
    for(let i = 0; i < getPointTypes().length; i++) {
        let pointType = getPointTypes()[i];
        for(let j = 0; j < pointType.length; j++) {
            if(pointType[j] === type) {
                return pointType;
            }
        }
    }
    return null;
}

function getBonusType(name) {
    for(let i = 0; i < getBonusTypes().length; i++) {
        let bonus = getBonusTypes()[i];
        for(let j = 0; j < bonus.aliases.length; j++) {
            if(bonus.aliases[j] === name) {
                return bonus;
            }
        }
    }
    return null;
}
